README.md
# mosquitto-stats

Send [Mosquitto](http://mosquitto.org) stats to Graphite/InfluxDB.


debian/changelog
mosquitto-stats (0.1-0rbel1) trusty; urgency=medium

  * Initial release.

 -- Sergio Rubio <rubiojr@frameos.org>  Wed, 18 Feb 2015 21:27:22 +0100


build.sh
#!/bin/bash
set -e

NAME=mosquitto-stats
VERSION=$(grep PKG_VERSION version.go | perl -pe 's|.*"([\d\.]+)".*|\1|')

BUILD_DIR=$HOME/debian/tmp/$NAME
PKG_NAME=$NAME-$VERSION
PKG=$BUILD_DIR/$PKG_NAME.tar.gz
DEB_TARGET_DIR=$HOME/debian/$NAME
ORIG_TARBALL=$DEB_TARGET_DIR/${NAME}_$VERSION.orig.tar.gz

orig_tarball() {
  rm -rf $BUILD_DIR && mkdir -p $BUILD_DIR
	git archive --output=$PKG --prefix=$PKG_NAME/ HEAD
	mv $PKG $BUILD_DIR/${NAME}_$VERSION.orig.tar.gz
}

srcdeb() {
  orig_tarball

	mkdir -p $DEB_TARGET_DIR
	# Prevents overwriting a good tarball
  if [ -f "$ORIG_TARBALL" ] && [ "$ARG1" != "--force" ]; then
    echo "Original tarball previously created, use --force to overwrite"
    exit 1
  fi

	cd $BUILD_DIR && \
		tar xzf ${NAME}_$VERSION.orig.tar.gz && \
	  cd $NAME-$VERSION && \
		debuild -S && rm -rf $BUILD_DIR/$NAME-$VERSION && \
		mv $BUILD_DIR/* $DEB_TARGET_DIR

  rm -rf "$BUILD_DIR"
}

CMD=$1
ARG1=$2
echo Building version $VERSION
case $CMD in
  srcdeb)
    srcdeb
    ;;
  deps)
    rm -rf vendor/src
    GOBIN=$PWD/vendor/bin GOPATH=$PWD/vendor go get
    rm -rf vendor/pkg
    rm -rf vendor/bin
    find vendor/src -type d -name '.git' -o -name '.hg' | xargs rm -rf
    find vendor/src -type f -name '.gitignore' -o -name '.hgignore' | xargs rm
    ;;
  format)
    gofmt -s -w **/*.go
    ;;
  clean)
    rm -f $NAME
    ;;
  *)
	  GOPATH=$PWD/vendor go build -o $NAME
    ;;
esac


mqttc/client.go
package mqttc

import (
	"crypto/tls"
	"crypto/x509"
	"errors"
	"fmt"
	"git.eclipse.org/gitroot/paho/org.eclipse.paho.mqtt.golang.git"
	log "github.com/Sirupsen/logrus"
	"io/ioutil"
	"net/url"
	"os"
	"regexp"
	"time"
)

type Args struct {
	Topic         string
	ClientID      string
	BrokerURLs    []string
	TLSCACertPath string
	TLSSkipVerify bool
}

// tcp://user:password@host:port
func PushMsg(msg string, args *Args) error {

	opts := mqtt.NewClientOptions()
	opts.SetCleanSession(true)
	opts.SetWriteTimeout(10 * time.Second)

	opts.SetClientId(args.ClientID)
	for _, broker := range args.BrokerURLs {
		uri, err := url.Parse(broker)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error parsing broker url (ignored): %s\n", broker)
			continue
		}
		if uri.Scheme == "ssl" {
			tlsconfig := newTlsConfig(args.TLSCACertPath, args.TLSSkipVerify)
			if !isTLSOK(*uri, tlsconfig) {
				continue
			}
			opts.SetTlsConfig(tlsconfig)
		}
		opts.AddBroker(broker)
	}

	client := mqtt.NewClient(opts)
	_, err := client.Start()
	if err != nil {
		return errors.New("Connection to the broker(s) failed: " + err.Error())
	}
	defer client.Disconnect(0)

	<-client.Publish(mqtt.QOS_ONE, args.Topic, msg)

	return nil
}

func Subscribe(handler mqtt.MessageHandler, args *Args) *mqtt.MqttClient {
	opts := mqtt.NewClientOptions()

	for _, broker := range args.BrokerURLs {
		uri, err := url.Parse(broker)
		if err != nil {
			log.Errorf("Error parsing broker url (ignored): %s\n", broker)
			continue
		}
		if uri.Scheme == "ssl" {
			tlsconfig := newTlsConfig(args.TLSCACertPath, args.TLSSkipVerify)
			if !isTLSOK(*uri, tlsconfig) {
				continue
			}
			opts.SetTlsConfig(tlsconfig)
		}
		log.Debug("Adding broker", broker)
		opts.AddBroker(broker)
	}

	opts.SetClientId(args.ClientID)
	opts.SetDefaultPublishHandler(handler)

	client := mqtt.NewClient(opts)
	_, err := client.Start()
	if err != nil {
		log.Fatal("Connection to the broker(s) failed: " + err.Error())
	}

	filter, e := mqtt.NewTopicFilter(args.Topic, byte(mqtt.QOS_ZERO))
	if e != nil {
		log.Panic(e)
	}

	if rectp, err := client.StartSubscription(nil, filter); err != nil {
		log.Panic(err)
	} else {
		<-rectp
	}

  log.Debug("Subscription to the brokers started")
	return client
}

func newTlsConfig(cacertPath string, verify bool) *tls.Config {
	if cacertPath == "" {
		return &tls.Config{}
	}

	certpool := x509.NewCertPool()
	pemCerts, err := ioutil.ReadFile(cacertPath)
	if err != nil {
		panic("Error reading CA certificate from " + cacertPath)
	}

	certpool.AppendCertsFromPEM(pemCerts)

	return &tls.Config{
		RootCAs:            certpool,
		InsecureSkipVerify: verify,
	}
}

// Test SSL connections to the brokers because the current
// paho mqtt client implementation returns a generic error message
// hard to debug.
func isTLSOK(uri url.URL, config *tls.Config) bool {
	r := regexp.MustCompile(".*:.*")
	if !r.MatchString(uri.Host) {
		uri.Host += ":8883"
	}

	_, err := tls.Dial("tcp", uri.Host, config)

	if err != nil {
		log.Warnf("Ignoring broker %s: %s", uri.String(), err)
		return false
	}

	return true
}


mosquitto-stats.go
package main

import (
	"./mqttc"
	"fmt"
	"git.eclipse.org/gitroot/paho/org.eclipse.paho.mqtt.golang.git"
	log "github.com/Sirupsen/logrus"
	"github.com/marpaia/graphite-golang"
	"gopkg.in/alecthomas/kingpin.v1"
	"net/url"
	"os"
	"strings"
	"time"
)

var (
	Graphite   *graphite.Graphite
	clientsMap map[*mqtt.MqttClient]*string
)

func sendMetrics(client *mqtt.MqttClient, msg mqtt.Message) {
	var host *string
	var ok bool
	if host, ok = clientsMap[client]; !ok {
		log.Warn("Not all the mqtt clients are ready")
		return
	}

	metric := normalizeMetric(msg.Topic())
	switch {
	case strings.Contains(metric, "version"):
		log.Debug("Skipping mosquitto version metric")
		return
	case strings.Contains(metric, "timestamp"):
		log.Debug("Skipping timestamp metric")
		return
	}

	value := string(msg.Payload())
	// mosquitto uptime metric appends 'seconds' to the number
	// we only need the value
	if metric == "mqtt.uptime" {
		value = strings.Split(string(msg.Payload()), " ")[0]
	}
	hostMetric := fmt.Sprintf("%s.%s", *host, metric)

	log.Debugf("Sending metric %s %s", hostMetric, value)
	err := Graphite.SimpleSend(hostMetric, value)
	if err != nil {
		log.Debugf("Error sending metric %s: %s", hostMetric, err)
	}
}

func parseBrokerUrls(brokerUrls string) []string {
	tokens := strings.Split(brokerUrls, ",")
	for i, url := range tokens {
		tokens[i] = strings.TrimSpace(url)
	}

	return tokens
}

func normalizeMetric(name string) string {
	m := strings.Replace(name, "$SYS/broker/", "mqtt.", -1)
	m = strings.Replace(m, "/", ".", -1)
	m = strings.Replace(m, " ", "_", -1)

	return m
}

func main() {
	kingpin.Version(PKG_VERSION)

	brokerUrls := kingpin.Flag("broker-urls", "Comman separated MQTT broker URLs").
		Required().Default("").OverrideDefaultFromEnvar("MQTT_URLS").String()

	cafile := kingpin.Flag("cafile", "CA certificate when using TLS (optional)").
		String()

	graphiteHost := kingpin.Flag("graphiteHost", "Graphite host").
		Default("localhost").String()

	graphitePort := kingpin.Flag("graphitePort", "Graphite port").
		Default("2003").Int()

	graphitePing := kingpin.Flag("graphitePing", "Try to reconnect to graphite every X seconds").
		Default("15").Int()

	insecure := kingpin.Flag("insecure", "Don't verify the server's certificate chain and host name.").
		Default("false").Bool()

	debug := kingpin.Flag("debug", "Print debugging messages").
		Default("false").Bool()

	kingpin.Parse()

	if *debug {
		log.SetLevel(log.DebugLevel)
	}

	if *cafile != "" {
		if _, err := os.Stat(*cafile); err != nil {
			log.Fatalf("Error reading CA certificate %s", err.Error())
			os.Exit(1)
		}
	}

	urlList := parseBrokerUrls(*brokerUrls)
	clientsMap = map[*mqtt.MqttClient]*string{}

	var gerr error
	Graphite, gerr = graphite.NewGraphite(*graphiteHost, *graphitePort)
	if gerr != nil {
		log.Warn("Error connecting to graphite")
		os.Exit(1)
	} else {
		log.Info("Connected to graphite")
		log.Debugf("Loaded Graphite connection: %#v", Graphite)
	}

	for _, urlStr := range urlList {
		args := mqttc.Args{
			BrokerURLs:    []string{urlStr},
			ClientID:      "mqtt-stats",
			Topic:         "$SYS/broker/#",
			TLSCACertPath: *cafile,
			TLSSkipVerify: *insecure,
		}

		uri, _ := url.Parse(urlStr)
		c := mqttc.Subscribe(sendMetrics, &args)
		defer c.Disconnect(0)
		host := strings.Split(uri.Host, ":")[0]
		host = strings.Replace(host, ".", "_", -1)
		clientsMap[c] = &host
	}

	// Try to reconnect every graphitePing sec if sending fails by sending
	// a fake ping metric
	// FIXME: better handled by the Graphite client
	go func() {
		for {
			time.Sleep(time.Duration(*graphitePing) * time.Second)
			err := Graphite.SimpleSend("ping metric", "")
			if err != nil {
				log.Warn("Ping metric failed, trying to reconnect")
				err = Graphite.Connect()
				if err != nil {
					log.Error("Reconnecting to graphite failed")
				} else {
					log.Info("Reconnected to graphite")
				}
			}
		}
	}()

	// wait endlessly
	var loop chan bool
	loop <- true
}


main.go
package main

import (
	"./mqttc"
	"fmt"
	"git.eclipse.org/gitroot/paho/org.eclipse.paho.mqtt.golang.git"
	log "github.com/Sirupsen/logrus"
	"github.com/marpaia/graphite-golang"
	"gopkg.in/alecthomas/kingpin.v1"
	"net/url"
	"os"
	"strings"
	"time"
)

var (
	Graphite   *graphite.Graphite
	clientsMap map[*mqtt.MqttClient]*string
)

func sendMetrics(client *mqtt.MqttClient, msg mqtt.Message) {
	var host *string
	var ok bool
	if host, ok = clientsMap[client]; !ok {
		log.Warn("Not all the mqtt clients are ready")
		return
	}

	metric := normalizeMetric(msg.Topic())
	switch {
	case strings.Contains(metric, "version"):
		log.Debug("Skipping mosquitto version metric")
		return
	case strings.Contains(metric, "timestamp"):
		log.Debug("Skipping timestamp metric")
		return
	}

	value := string(msg.Payload())
	// mosquitto uptime metric appends 'seconds' to the number
	// we only need the value
	if metric == "mqtt.uptime" {
		value = strings.Split(string(msg.Payload()), " ")[0]
	}
	hostMetric := fmt.Sprintf("%s.%s", *host, metric)

	log.Debugf("Sending metric %s %s", hostMetric, value)
	err := Graphite.SimpleSend(hostMetric, value)
	if err != nil {
		log.Debugf("Error sending metric %s: %s", hostMetric, err)
	}
}

func parseBrokerUrls(brokerUrls string) []string {
	tokens := strings.Split(brokerUrls, ",")
	for i, url := range tokens {
		tokens[i] = strings.TrimSpace(url)
	}

	return tokens
}

func normalizeMetric(name string) string {
	m := strings.Replace(name, "$SYS/broker/", "mqtt.", -1)
	m = strings.Replace(m, "/", ".", -1)
	m = strings.Replace(m, " ", "_", -1)

	return m
}

func main() {
	kingpin.Version("0.1")

	brokerUrls := kingpin.Flag("broker-urls", "Comman separated MQTT broker URLs").
		Required().Default("").OverrideDefaultFromEnvar("MQTT_URLS").String()

	cafile := kingpin.Flag("cafile", "CA certificate when using TLS (optional)").
		String()

	graphiteHost := kingpin.Flag("graphiteHost", "Graphite host").
		Default("localhost").String()

	graphitePort := kingpin.Flag("graphitePort", "Graphite port").
		Default("2003").Int()

	graphitePing := kingpin.Flag("graphitePing", "Try to reconnect to graphite every X seconds").
		Default("15").Int()

	insecure := kingpin.Flag("insecure", "Don't verify the server's certificate chain and host name.").
		Default("false").Bool()

	debug := kingpin.Flag("debug", "Print debugging messages").
		Default("false").Bool()

	kingpin.Parse()

	if *debug {
		log.SetLevel(log.DebugLevel)
	}

	if *cafile != "" {
		if _, err := os.Stat(*cafile); err != nil {
			log.Fatalf("Error reading CA certificate %s", err.Error())
			os.Exit(1)
		}
	}

	urlList := parseBrokerUrls(*brokerUrls)
	clientsMap = map[*mqtt.MqttClient]*string{}

	var gerr error
	Graphite, gerr = graphite.NewGraphite(*graphiteHost, *graphitePort)
	if gerr != nil {
		log.Warn("Error connecting to graphite")
		os.Exit(1)
	} else {
		log.Info("Connected to graphite")
		log.Debugf("Loaded Graphite connection: %#v", Graphite)
	}

	for _, urlStr := range urlList {
		args := mqttc.Args{
			BrokerURLs:    []string{urlStr},
			ClientID:      "mqtt-stats",
			Topic:         "$SYS/broker/#",
			TLSCACertPath: *cafile,
			TLSSkipVerify: *insecure,
		}

		uri, _ := url.Parse(urlStr)
		c := mqttc.Subscribe(sendMetrics, &args)
		defer c.Disconnect(0)
		host := strings.Split(uri.Host, ":")[0]
		host = strings.Replace(host, ".", "_", -1)
		clientsMap[c] = &host
	}

	// Try to reconnect every graphitePing sec if sending fails by sending
	// a fake ping metric
	// FIXME: better handled by the Graphite client
	go func() {
		for {
			time.Sleep(time.Duration(*graphitePing) * time.Second)
			err := Graphite.SimpleSend("ping metric", "")
			if err != nil {
				log.Warn("Ping metric failed, trying to reconnect")
				err = Graphite.Connect()
				if err != nil {
					log.Error("Reconnecting to graphite failed")
				} else {
					log.Info("Reconnected to graphite")
				}
			}
		}
	}()

	// wait endlessly
	var loop chan bool
	loop <- true
}
